### PR TITLE
VideoCommon: Start refactoring TextureCacheBase::Load

### PR DESCRIPTION
**Pull request by hthh**
I think this is a step in the right direction (and won't break anything).

This fixes:
 * Duplicate use of `width` and `height` to mean "requested width" and "width to load (possibly from high res)"
 * Overlapping use of `texLevels` and `tex_levels` to mean different things
 * Duplicate logic invoking `TexDecoder_Decode`
 * A few other little things

It should also fixes RGBA mipmap loading from tmem, but this is (still) untested.

The new `TextureRequest` thing can probably be used in the software backend too, particularly if we add a cache for them.

### CODE REVIEW

**Comment by hthh** [comment]
@stenzek I fixed the graphical regression (by not zeroing `m_address` for tmem loads 😕). Can I get your thoughts on this? Particularly with a view to any changes I could make to facilitate future texture cache changes (e.g. locking, copy elimination). For now I'm avoiding refactoring the caching itself - hopefully it'll be easier after this is out of the way.

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 1]
```diff
@@ -484,140 +483,205 @@ void TextureCacheBase::UnbindTextures()
...
+      m_num_levels = std::min<u32>(IntLog2(std::max(width, height)) + 1, levels);
+    }
+    else
+    {
+      m_num_levels = 1;
+    }
 
-  if (isPaletteTexture)
-    full_format = texformat | (tlutfmt << 16);
+    const u8* src_data = m_from_tmem ? nullptr : Memory::GetPointer(m_address);
```
Comment (iwubcode) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: Off topic - I'll admit, I don't usually const'ify most of my variables unless it's a pass in from a function.  What's the gain?
More on topic - maybe this optimization can already be done by the compiler.  But It seems odd we're branching 3 times to define some variables.  I almost wonder if the for loop could go into separate function called InitializeLevels().  Then you could do:

```
if (m_from_tmem)
{
  InitializeLevels(nullptr, &texMem[image1.tmem_even * TMEM_LINE_SIZE]), &texMem[image2.tmem_odd * TMEM_LINE_SIZE]);
}
else
{
  InitializeLevels(Memory::GetPointer(m_address), nullptr, nullptr);
}
```

Or potentially to mimic your approach with DecodeLevel, we have InitializeLevel and keep the for loop?

I don't feel very strongly about this though, just looking to improve code clarity and performance.

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 2]
```diff
@@ -731,35 +793,37 @@ TextureCacheBase::TCacheEntryBase* TextureCacheBase::Load(const u32 stage)
     InvalidateTexture(oldest_entry);
   }
 
+  // Loaded texture data (which may not match requested texture data)
+  u32 width = request.Width();
+  u32 height = request.Height();
+  u32 stride = request.m_levels[0].padded_width;
+
   std::shared_ptr<HiresTexture> hires_tex;
   if (g_ActiveConfig.bHiresTextures)
```
Comment (iwubcode) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: One thing about this is I really would like to think we could improve the clarity here...there's so much hopping around based on if we're dealing with highres or "lowres" textures.  Could we instead break it down to something like this:

```
if (g_ActiveConfig.bHiresTextures)
{
  CalculateHiResTextures(entry);
}
else
{
  CalculateLowResTextures(entry);
}

return ReturnEntry(stage, entry);
```

Not sure what all parameters we would need to make this work.

**Comment by iwubcode** [comment]
@hthh - overall good job, I forgot you mentioned you were working on this.  I actually looked at it a couple days ago and thought it needed a rewrite.  I like what you're doing here!

File: Source/Core/VideoCommon/TextureCacheBase.cpp (hunk 1 above)
Comment (hthh) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: `const` has a few benefits - I don't typically use it to this extent in my own code, but prior to my changes the function had `const` on most variables, notably excluding `width` and `height` which were, counterintuitively, non-constant. The use of `const` made it a lot easier to reason about that code and see where tricky things were happening.

The use of `const u8*` makes it clear that we aren't modifying the texture data, and will make it clear that it's safe to change this code if we introduce a `const u8 *Memory::GetReadPointer(u32 addr, u32 size)` function in the future (which we are quite likely to want to do for locking, with write-protected underlying memory).

There's no point optimising this code for branch count. If it's a bottleneck, it'd be much more effective to make some of the calculations lazy, or reduce the divisions in mipmap count. (And if we can clean up texture caching and a few other things, we should be able to implement zero-copy uploads and locking, which would be (relatively) huge performance wins). Sorry if that's a bit of ramble, just hoping to explain why I'm much more interested in improving clarity than immeasurable performance improvements.

I'll investigate separating out an `InitializeLevels` function and see if it makes things clearer.

Comment (Parlane) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: Using const for local vars is now recommended for all Dolphin code.  It explicitly says "do not change this" and the compiler can be smarter about how it loads them in to memory.

Comment (iwubcode) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: Thanks to you both!  I did know the primary reason was enforcing constantness of the variable but I was curious if there was a more subtle reason const was used all over the place in Dolphin code.

In particular, I wouldn't have used it due to the branching three times but hthh makes a good point that maybe performance loss is negligible.  I probably prioritize minimizing branching more than I should.

**Comment by nbohr1more** [comment]
Rebase please?

**Comment by jordan-woyak** [comment]
I'm closing this PR for no rebase after 3 years. Will reopen upon request.

**Pull request closed without merging**